### ISSUE TITLE
Generator expressions are type checked as if they are immediately executed

### ISSUE THREAD

**Issue by abextm** [microsoft/pyright#5376]
**Describe the bug**
The target expression of a generator expression is type checked assuming any variables it closes over never change type again, even though this may not be true. Pyright accepts the following code, even though it fails at runtime, as iterating over the generator changes the type of `path` from from `Literal["foo"]` to `List[str]`.

```python
# pyright: strict
path = "foo"
entries = ([path + "/" +  name] for name in ["1", "2", "3"])
for path in entries:
  print(path)
```

It even accepts this, which trips the assert at runtime
```python
# pyright: strict

from typing import Literal
def take_foo(a: Literal["foo"]) -> str:
        assert a == "foo"
        return a

path = "foo"
entries = ([take_foo(path) + "/" +  name] for name in ["1", "2", "3"])
for path in entries:
  print(path)
```

if you make the expression `lambda: take_foo(path) + "/" +  name`, Pyright will correctly complain that `path` changes types

**To Reproduce**
Execute Pyright on the above snippets. It will output `0 errors, 0 warnings, 0 informations`

**Expected behavior**
Pyright should treat the target of a generator expression like it does a lambda, and complain when it changes types

**VS Code extension or command-line**
```
$ pyright --version
pyright 1.1.315
```

**Comment by erictraut** [comment]
Type checkers necessarily need to make certain assumptions to strike the right balance between false positives and false negatives. Your code violates an assumption that pyright makes about generators. If it doesn't make this assumption, I fear it will generate many false positive errors in common code patterns. That wouldn't be a good tradeoff.

I'll think about this more and experiment to see if I can come up with something, but my initial thought is that the current behavior strikes the right balance.

**Comment by abextm** [comment]
While I agree that this is more restrictive than it used to be, opting into that restriction is the whole point of a type checker. Its not uncommon to run into this sort of thing in typescript, where it is a failure. It can be annoying, but it also prevents bugs which is (imo) more important. Not catching incorrect code makes pyright significantly less useful to me.

**Comment by erictraut** [comment]
Yes, I understand your perspective. There will always be some bugs that a static type checker will not catch. If a type checker generates false positive errors in too many cases, its utility will likewise be diminished for most users. Striking the right balance between false positives and false negatives is something we continually think about when developing pyright. The TypeScript team also thinks about this a great deal, and we tend to make similar tradeoffs.

**Comment by erictraut** [comment]
I implemented the change and run it on a bunch of large code bases that use pyright. As I predicted, it generated a bunch of new false positive errors and found no new legitimate bugs. So I don't think this is a bit of an anti-feature.

What you're doing in your code above is very atypical, so I don't think it's a problem that pyright doesn't catch this bug. Mypy also doesn't catch the issue.

You can avoid the problem if you provide an explicit type annotation for `path`.

```python
path: str = "foo"
```

I'm going to close the feature request for now. I would be willing to revisit it in the future if there's sufficient demand from pyright users, but I would be somewhat surprised if there was.
